server/src/test/java/org/elasticsearch/cluster/routing/BatchedRerouteServiceTests.java

```diff
@@ -229,6 +235,93 @@ public void testNotifiesOnFailure() throws InterruptedException {
+                .setClusterStatePublisher(
+                    (event, publishListener, ackListener) -> publishListener.onFailure(new FailedToCommitClusterStateException("simulated"))
+                );
+
+            mockLogAppender.addExpectation(
+                new MockLogAppender.SeenEventExpectation(
+                    "unexpected failure",
+                    BatchedRerouteService.class.getCanonicalName(),
+                    Level.ERROR,
+                    "unexpected failure"
+                )
+            );
+
+            final BatchedRerouteService failingRerouteService = new BatchedRerouteService(clusterService, (s, r, l) -> {
+                throw new ElasticsearchException("simulated");
+            });
+            final var rerouteFailureFuture = new PlainActionFuture<Void>();
+            failingRerouteService.reroute("publish failure", randomFrom(EnumSet.allOf(Priority.class)), rerouteFailureFuture);
+            assertThat(
+                expectThrows(ExecutionException.class, ElasticsearchException.class, () -> rerouteFailureFuture.get(10, TimeUnit.SECONDS))
+                    .getMessage(),
+                equalTo("simulated")
+            );
+            mockLogAppender.assertAllExpectationsMatched();
+
+            final BatchedRerouteService batchedRerouteService = new BatchedRerouteService(clusterService, (s, r, l) -> {
+                l.onResponse(null);
+                return ClusterState.builder(s).build();
+            });
+
+            mockLogAppender.addExpectation(
+                new MockLogAppender.UnseenEventExpectation("no errors", BatchedRerouteService.class.getCanonicalName(), Level.ERROR, "*")
+            );
+            mockLogAppender.addExpectation(
+                new MockLogAppender.UnseenEventExpectation("no warnings", BatchedRerouteService.class.getCanonicalName(), Level.WARN, "*")
+            );
+            mockLogAppender.addExpectation(
+                new MockLogAppender.UnseenEventExpectation("no info", BatchedRerouteService.class.getCanonicalName(), Level.INFO, "*")
+            );
+            mockLogAppender.addExpectation(
+                new MockLogAppender.SeenEventExpectation(
+                    "publish failure",
+                    BatchedRerouteService.class.getCanonicalName(),
+                    Level.DEBUG,
+                    "unexpected failure"
+                )
+            );
+
+            final var publishFailureFuture = new PlainActionFuture<Void>();
+            batchedRerouteService.reroute("publish failure", randomFrom(EnumSet.allOf(Priority.class)), publishFailureFuture);
+            expectThrows(
+                ExecutionException.class,
+                FailedToCommitClusterStateException.class,
+                () -> publishFailureFuture.get(10, TimeUnit.SECONDS)
+            );
+            mockLogAppender.assertAllExpectationsMatched();
+
+            PlainActionFuture.<Void, RuntimeException>get(future -> {
```

Comment:
`// Third case: assert a non-master failure results in an unwrapped NotMasterException`